[SPARK-9483] Fix UTF8String.getPrefix for big-endian.

mtbrandy opened the pull request:
Previous code assumed little-endian.

mtbrandy committed:
[SPARK-9483] Fix UTF8String.getPrefix for big-endian.

Previous code assumed little-endian.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

JoshRosen commented:
/cc @davies

JoshRosen commented:
Jenkins, this is ok to test.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:
```diff
@@ -161,18 +162,25 @@ public long getPrefix() {
+      mask = (1L << (8 - numBytes) * 8) - 1;
       p = (long) PlatformDependent.UNSAFE.getInt(base, offset);
-      p = p & ((1L << numBytes * 8) - 1);
+      if (ByteOrder.nativeOrder() == ByteOrder.BIG_ENDIAN) {
```
can we move the result of ByteOrder.nativeOrder into a static variable?

mtbrandy replied:
Will do

SparkQA commented:
[Test build #39580 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39580/consoleFull) for   PR 7902 at commit [`17d54c6`](https://github.com/apache/spark/commit/17d54c6bb898e27cbb386c214921ad4feaf106f8).

rxin commented on unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java:
```diff
@@ -161,18 +162,25 @@ public long getPrefix() {
     long p;
+    long mask = 0;
     if (numBytes >= 8) {
       p = PlatformDependent.UNSAFE.getLong(base, offset);
```
don't we need something for this branch also?

rxin replied:
ah nvm u don't need it since the reversebytes was done later.

I think it's not that big of a deal to duplicate code here, and the nested ifs really make it harder to read.

mtbrandy replied:
The mask represents the characters that need to be cleared. p needs no modification (and thus mask of zero is correct) when numBytes is equal to zero or greater than 8.

rxin commented:
@mtbrandy  not 100% sure, but it might be more clear if you just have a top level branch based on endianness.

mtbrandy commented:
@rxin, regarding top-level branch -- I can do that, but was just trying to minimize duplicate code.

davies commented:
+1 for moving the endianness check to top level, otherwise LGTM.

mtbrandy commented:
ok, I'll modify to move the endianness check up.

mtbrandy committed:
[SPARK-9483] Changes from review comments.

- Introduce static byteOrder variable
- Branch on byteOrder at top-level for readability.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39589 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39589/consoleFull) for   PR 7902 at commit [`ec31df8`](https://github.com/apache/spark/commit/ec31df8ce5b32b747224212bef15e3f85d96a572).

SparkQA commented:
[Test build #39580 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39580/console) for   PR 7902 at commit [`17d54c6`](https://github.com/apache/spark/commit/17d54c6bb898e27cbb386c214921ad4feaf106f8).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

rxin commented:
@davies I will let you merge this.

SparkQA commented:
[Test build #39589 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39589/console) for   PR 7902 at commit [`ec31df8`](https://github.com/apache/spark/commit/ec31df8ce5b32b747224212bef15e3f85d96a572).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

davies commented:
Merged into master and 1.5-branch, thanks!
